EQL: support for DynamicEntityKey as value for query parameters

### Description

Currently, passing `DynamicEntityKey` values into `.val()` and `.param()` result in a runtime exception
```
Caused by: org.hibernate.HibernateException: Could not determine a type for class: ua.com.fielden.platform.entity.DynamicEntityKey
```
Similarly to supporting `DynamicEntityKey` values in search by key companion APIs, it would be convenient to support them in EQL by converting such values to string for parameter assignment. Such support could also help with refactoring of domain models, where `String` keyed entities become composite entities.

### Expected outcome
Support for `DynamicEntityKey` values as EQL parameters. Potential productivity gains from slightly richer support for composite entities in EQL.

--- a/platform-dao/src/test/java/ua/com/fielden/platform/dao/CommonEntityDaoTest.java
+++ b/platform-dao/src/test/java/ua/com/fielden/platform/dao/CommonEntityDaoTest.java
@@ -24,12 +24,12 @@ import java.util.Optional;
 import java.util.stream.Stream;
 
 import org.joda.time.DateTime;
-import org.junit.Ignore;
 import org.junit.Test;
 
 import ua.com.fielden.platform.dao.exceptions.EntityCompanionException;
 import ua.com.fielden.platform.entity.AbstractEntity;
 import ua.com.fielden.platform.entity.DynamicEntityKey;
+import ua.com.fielden.platform.entity.query.fluent.fetch;
 import ua.com.fielden.platform.entity.query.model.EntityResultQueryModel;
 import ua.com.fielden.platform.entity.query.model.OrderingModel;
 import ua.com.fielden.platform.pagination.IPage;
@@ -891,12 +891,43 @@ public class CommonEntityDaoTest extends AbstractDaoTestCase {
         co.save(entity);
     }
 
+    @Test
+    public void composite_entities_can_by_queried_by_passing_composite_key_values_as_EQL_parameters() {
+        final IEntityDao<EntityWithDynamicCompositeKey> co = co(EntityWithDynamicCompositeKey.class);
+        final String compositeKeyAsString = "key-1-1" + " " + "KEY1";
+        final EntityResultQueryModel<EntityWithDynamicCompositeKey> qByKeyAsString = select(EntityWithDynamicCompositeKey.class).where()
+                .prop("key").eq().val(compositeKeyAsString).model();
+        final fetch<EntityWithDynamicCompositeKey> fetch = fetchAll(EntityWithDynamicCompositeKey.class);
+        final EntityWithDynamicCompositeKey entityByKeyAsString = co.getEntity(from(qByKeyAsString).with(fetch).model());
+        assertNotNull(entityByKeyAsString);
+
+        final EntityResultQueryModel<EntityWithDynamicCompositeKey> qByKey = select(EntityWithDynamicCompositeKey.class).where()
+                .prop("key").eq().val(entityByKeyAsString.getKey()).model();
+        final EntityWithDynamicCompositeKey entityByKey = co.getEntity(from(qByKey).with(fetch).model());
+        assertNotNull(entityByKey);
+        assertEquals(entityByKeyAsString, entityByKey);
+    }
+
+    @Override
+    public boolean saveDataPopulationScriptToFile() {
+        return false;
+    }
+
+    @Override
+    public boolean useSavedDataPopulationScript() {
+        return false;
+    }
+
     @Override
     protected void populateDomain() {
         super.populateDomain();
 
         final UniversalConstantsForTesting constants = (UniversalConstantsForTesting) getInstance(IUniversalConstants.class);
         constants.setNow(dateTime("2016-02-19 02:47:00"));
+
+        if (useSavedDataPopulationScript()) {
+            return;
+        }
 
         final EntityWithMoney keyPartTwo = save(new_(EntityWithMoney.class, "KEY1", "desc").setMoney(new Money("20.00")).setDateTimeProperty(date("2009-03-01 11:00:55")));
         save(new_composite(EntityWithDynamicCompositeKey.class, "key-1-1", keyPartTwo));

--- a/platform-pojo-bl/src/main/java/ua/com/fielden/platform/entity/query/fluent/ValuePreprocessor.java
+++ b/platform-pojo-bl/src/main/java/ua/com/fielden/platform/entity/query/fluent/ValuePreprocessor.java
@@ -10,12 +10,20 @@ import java.util.Collection;
 import java.util.List;
 
 import ua.com.fielden.platform.entity.AbstractEntity;
+import ua.com.fielden.platform.entity.DynamicEntityKey;
 import ua.com.fielden.platform.entity.meta.PropertyDescriptor;
 import ua.com.fielden.platform.types.Colour;
 import ua.com.fielden.platform.types.Hyperlink;
 import ua.com.fielden.platform.types.Money;
 
+/**
+ * This class is responsible for pre-processing of values passed as parameters to EQL queries (this covers both {@code .val()} and {@code .param()}).
+ *
+ * @author TG Team
+ *
+ */
 public class ValuePreprocessor {
+
     public Object apply(final Object value) {
         if (value == null) {
             return null;
@@ -38,7 +46,8 @@ public class ValuePreprocessor {
             value instanceof Class ||
             value instanceof Colour ||
             value instanceof Enum ||
-            value instanceof Hyperlink) {
+            value instanceof Hyperlink ||
+            value instanceof DynamicEntityKey) {
             result = value.toString();
         } else if (value instanceof AbstractEntity) {
             final AbstractEntity<?> entity = (AbstractEntity<?>) value;

